lib/internal/process/promises.js

```diff
@@ -60,11 +58,23 @@ function setupPromises(scheduleMicrotasks) {
   }
 
   function emitWarning(uid, reason) {
+    try {
+      if (reason instanceof Error) {
+        process.emitWarning(reason.stack, 'UnhandledPromiseRejectionWarning');
+      } else {
+        process.emitWarning(reason, 'UnhandledPromiseRejectionWarning');
```

Comment:
So this still should be `safeToString(reason)`, notice that `emitWarning` will only emit strings:

Reply:
What's the expected behavior when strings get rejected? I couldn't find a way to get the stack trace in that case.

Reply:
I'd assume the stack has been destroyed when this is reached. IMHO just printing the reason (after stringifying) is the best we can do.
So if it's an Error `reason.stack` else `safeToString(reason)`.

Reply:
I think that `test-promises-warning-on-unhandled-rejection.js` implicitly checks for string rejections, please tell me if something else might be better.

I've pushed the commit to wrap the reason with `safeToString()`.

Reply:
Yeah it's an awkward test... You could add a line just before last:
```js
const p = Promise.reject(100);
```
then more case (2) to (3) add a case (2):
```
case 2:
      assert.strictEqual(warning.name, 'UnhandledPromiseRejectionWarning');
      assert(/100/.test(warning.message));
      break;
```

Reply:
Sorry, should be `p1` since `p` is used for the async handling check

Reply:
@refack Added the requested test, also added comments to make it clear what each state in the state machine means.

Reply:
💯